pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,623 @@
+			ifindex, err)
+	}
+
+	var (
+		stack revert.RevertStack
+
+		oldTable = rule.Table
+	)
+
+	revert, err := m.copyRoutes(routes, oldTable, newTable)
+	stack.Extend(revert)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new routes: %w", err)
+	}
+
+	revert, err = m.createNewRule(
+		rule,
+		linux_defaults.RulePriorityEgressv2,
+		newTable,
+	)
+	stack.Extend(revert)
+	if err != nil {
+		// We revert here because we want to ensure that the new routes
+		// are removed as they'd have no effect, but may conflict with
+		// others in the future.
+		if revErr := stack.Revert(); revErr != nil {
+			scopedLog.WithError(err).WithField("revertError", revErr).Warn(upgradeRevertWarning)
+		}
+
+		return nil, fmt.Errorf("failed to create new rule: %w", err)
+	}
+
+	if err := m.deleteOldRule(rule); err != nil {
+		// We revert here because we want to ensure that the new state
+		// that we just created above is reverted.
+		if revErr := stack.Revert(); revErr != nil {
+			scopedLog.WithError(err).WithField("revertError", revErr).Warn(upgradeRevertWarning)
+		}
+
+		return nil, fmt.Errorf("failed to delete old rule: %w", err)
+	}
+
+	return routes, nil
+}
+
+// downgradeRule migrates the given rule (and endpoint) to the original ENI
+// datapath, using the old table ID scheme that was simply the ifindex of the
+// attached ENI device on the node. It returns the "old" routes (new datapath)
+// that the caller should remove at a later time, along with an error.
+func (m *migrator) downgradeRule(rule netlink.Rule) ([]netlink.Route, error) {
+	scopedLog := log.WithField("rule", rule)
+
+	oldTable := rule.Table
+	ifaceNumber := oldTable - linux_defaults.RouteTableInterfacesOffset
+
+	newTable, err := m.retrieveTableIDFromInterfaceNumber(ifaceNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve new table ID from interface-number %q: %w",
+			ifaceNumber, err)
+	}
+
+	routes, err := m.rpdb.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: oldTable,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list routes associated with rule: %w", err)
+	}
+
+	var stack revert.RevertStack
+
+	revert, err := m.copyRoutes(routes, oldTable, newTable)
+	stack.Extend(revert)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new routes: %w", err)
+	}
+
+	revert, err = m.createNewRule(
+		rule,
+		linux_defaults.RulePriorityEgress,
+		newTable,
+	)
+	stack.Extend(revert)
+	if err != nil {
+		if revErr := stack.Revert(); revErr != nil {
+			scopedLog.WithError(err).WithField("revertError", revErr).Warn(downgradeRevertWarning)
+		}
+
+		return nil, fmt.Errorf("failed to create new rule: %w", err)
+	}
+
+	if err := m.deleteOldRule(rule); err != nil {
+		if revErr := stack.Revert(); revErr != nil {
+			scopedLog.WithError(err).WithField("revertError", revErr).Warn(downgradeRevertWarning)
+		}
+
+		return nil, fmt.Errorf("failed to delete old rule: %w", err)
```

Comment:
I think that this should actually 
```suggestion
		return nil, nil
```

Since the rule insertion above actually made the downgrade effective. Thoughts?

Reply:
^^^ Something I missed in my original post here, if so we should also remove the RevertStack handling logic a few lines above.

Reply:
Yes, that's correct. Will fix.

Reply:
On second thought, would it be useful for the user to be informed of this, so that they can cleanup the leftover state themselves?

Reply:
Yes, I think we should log like we do for the other "failed cleanup" steps.